### PR TITLE
Separate compute streams in forward path

### PR DESCRIPTION
**Pull request by moonbucks**
## Description

Separate streams in forward path and allows concurrent execution

## Type of change

Please delete options that are not relevant.

- [v] Bug fix (non-breaking change which fixes an issue)
- [v] New feature (non-breaking change which adds functionality)

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [v] Test LocalTestForwardTest
equivalence test passed 196872352.0 ref 196872352.0
profiling run completed 196872352.0 ref 196872352.0

- [v] Test LocalTestTwostreamTest
passed

- [v] Test LocalTestForwardBackwardTest
equivalence test passed 2.833994441818112e+16 ref 2.833994227069747e+16
Gradient equivalence test passed
correctness checks with original module passed

## Checklist:

- [v] Have you added tests that prove your fix is effective or that this feature works?
- [v] Has code been commented, particularly in hard-to-understand areas?
- [v] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: pippy/PipelineStage.py [hunk 1]
```diff
@@ -70,6 +70,9 @@ def __init__(
         self.kwargs_chunk_spec = kwargs_chunk_spec
         self.output_chunk_spec = output_chunk_spec
 
+        self.even = torch.cuda.Stream()
+        self.odd = torch.cuda.Stream()
```
Comment (fduwjj) [inline pippy/PipelineStage.py:74]: instead of hard-code even and odd, would it be possible that we let user to set number of streams to use and we have a list of streams?

File: test/local_test_twostream.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        x = self.mlp1(x)
+        pipe_split()
+        x = self.mlp2(x)
+        pipe_split()
+        x = self.mlp3(x)
+        loss = self.mse_loss(x, target)
+        return {"logits": x, "loss": loss}
+
+
+class ExampleCodeRef(torch.nn.Module):
```
Comment (fduwjj) [inline test/local_test_twostream.py:53]: Looks like this class is not even used?

Comment (moonbucks) [reply test/local_test_twostream.py:53]: oh.. I added it for initial test but forgot to remove it. fixed in the followed commit

File: pippy/PipelineStage.py (hunk 1 above)
Comment (moonbucks) [reply pippy/PipelineStage.py]: we can make it to have a stream for each chunk. please see the following commit

File: test/local_test_twostream.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    example_input = torch.randn(
+        args.batch_size, args.hidden_size, device=args.device
+    )
+    example_target = torch.randn(
+        args.batch_size, args.hidden_size, device=args.device
+    )
+
+    model.to(args.device)
+    ref_out = model(example_input, example_target)
```
Comment (fduwjj) [inline test/local_test_twostream.py:121]: not a blocker, but we have not tested bwd yet right?

File: pippy/PipelineStage.py [hunk 4]
```diff
@@ -70,6 +70,10 @@ def __init__(
         self.kwargs_chunk_spec = kwargs_chunk_spec
         self.output_chunk_spec = output_chunk_spec
 
+        self.streams = []
+        for i in range(chunks):
+            self.streams.append(torch.cuda.Stream())
```
Comment (fduwjj) [inline pippy/PipelineStage.py:75]: this looks good to me, but originally I mean we can let users set how many streams they want to use.  We can let them set 2/3 or maybe like what you are doing here # of stages. Basically this number falls in the range of [2, # of stages]. But we can address this in a separate PR.

**Review by fduwjj — APPROVED** [review]
unblock and please make sure CI passed.

File: test/local_test_twostream.py (hunk 3 above)
Comment (moonbucks) [reply test/local_test_twostream.py:121]: pippy originally has backward test and this modification passed the backward test. This feature didn't modify the behavior ( stream or synchronization) of backward path so I didn't add the test.

File: pippy/PipelineStage.py (hunk 4 above)
Comment (moonbucks) [reply pippy/PipelineStage.py:75]: resolved in a subsequent PR: https://github.com/pytorch/PiPPy/pull/852
